ds_exploration/data_download.py

```diff
@@ -8,79 +8,100 @@
-        array = rasterio.open(band_file).read(1)
-        return array
+        return rasterio.open(band_file).read(1)
 
-    red = get_band_array("B04", band_files)
-    green = get_band_array("B03", band_files)
-    blue = get_band_array("B02", band_files)
-    b12 = get_band_array("B12", band_files)
+    red, green, blue = [get_band_array(band, band_files) for band in ['B04', 'B03', 'B02']]
+    b12 = get_band_array('B12', band_files)
     return (np.dstack([red, green, blue]), b12)
 
+
+def save_originals(file_path, output_folder, tile_title, keep, delete_unused):
+    Path(file_path).mkdir(exist_ok=True)
+    band_files = glob(str(Path(f'{file_path}.SAFE/GRANULE/*/IMG_DATA/*.jp2')))
+    to_move = [file for file in band_files if any(file.endswith(f'{band}.jp2') for band in keep)]
+    for file in to_move:
+        Path(file).replace(Path(f'{output_folder}/{tile_title}/{Path(file).name}'))
+    if delete_unused:
+        shutil.rmtree(Path(f'{file_path}.SAFE'))
+
+
+def query_and_download(tile, api, cols, output_folder, nodefilter):
+    print(f'Download tile name: {tile["tile_name"]} for {tile["date"][0]}, {tile["date"][1]}, with filter {nodefilter}')
+    name = tile['tile_name']
+    date = tile['date']
+    products = api.query(date=date, tileid=name, platformname='Sentinel-2', producttype='S2MSI1C')
+    
+    downloaded_products = (
+        api.download_all(products, directory_path=output_folder, n_concurrent_dl=5, nodefilter=nodefilter)
+        if nodefilter else
+        api.download_all(products, directory_path=output_folder, n_concurrent_dl=5)
+    )
+
+    return pd.DataFrame([
+        [v[col] for col in cols] for (_, v) in downloaded_products[0].items()
+    ], columns=cols).set_index('id')
+
+
+def extract_data_from_tile(tile, output_folder, delete_unused):
+    file_path = Path(f'{output_folder}/{tile.title}')
+    unarchive_download(file_path)
+    rgb, b12 = extract_and_combine(file_path)
+    cut_and_save(file_path, rgb, tile.title)
+    save_originals(file_path, output_folder, tile.title, ['B04', 'B03', 'B02', 'B12'], delete_unused)
+    return rgb, b12
+
+def process_products(products, output_folder, delete_unused):
+    print(f'Downloaded {len(products)} tiles, extracting and combining data')
+    layers = []
+    for tile in products.itertuples():
+        layers.append(extract_data_from_tile(tile, output_folder, delete_unused))
+    return pd.DataFrame.from_records(layers, columns=['rgb', 'b12'], index=products.index)
+
+
 def download_tiles(list_of_tiles,
                    output_folder,
                    user,
                    password,
-                   nfilter='*B??.jp2',
+                   delete_unused=True,
+                   nodefilter=None,
                   ):
-    try:
-        os.mkdir(output_folder)
-    except:
-        pass
+
+    Path(output_folder).mkdir(exist_ok=True)
     
-    if nfilter:
-        nodefilter=make_filter(nfilter)
+    if nodefilter:
```

Comment:
Since this is not used atm, shall we remove it?

Reply:
Yeah, I still wanted to give it a try and fix the problem to make it work, but I guess we could just park somewhere that part of the code at the moment.